packages/victory-core/src/victory-util/hooks/use-domain.ts

```diff
@@ -0,0 +1,105 @@
+import * as React from "react";
+import * as d3Array from "victory-vendor/d3-array";
+import { PaddingProps } from "../../victory-theme/victory-theme";
+
+// TODO: Move this to ../types.ts
+type ForAxes<T> = T | { x?: T; y?: T };
+type Tuple<T> = [T, T];
+type Axis = "x" | "y";
+type ValueOrAxesObject<T> = T | ForAxes<T>;
+
+// TODO: There may be other types of data
+type Datum = number | Date;
+type Dataset = { x: Datum; y: Datum }[];
+type DomainTuple = Tuple<Datum>;
+
+interface DomainProps {
+  domain?: ValueOrAxesObject<DomainTuple>;
+  maxDomain?: ValueOrAxesObject<number>;
+  minDomain?: ValueOrAxesObject<number>;
+  domainPadding?: ValueOrAxesObject<number | Tuple<number>>;
+  width?: number;
+  height?: number;
+  padding?: PaddingProps;
+  // This should be a flattened list of all chart data.
+  // This data should be cleaned and formatted at the top-level
+  // before being passed into this component. Previously there
+  // was logic for converting this data back from preformatted
+  // data, and it would be better to standardize the data format
+  // and share the same formatted data between all modules.
+  data?: Dataset;
+}
+
+function hasValueForAxis<T = unknown>(
+  value: unknown | ForAxes<T>,
+  axis: Axis
+): value is ForAxes<T> {
+  if (typeof value === "object" && value !== null) {
+    return axis in value;
+  }
+  return false;
+}
+
+function isTuple<T = unknown>(value: unknown): value is Tuple<T> {
+  return Array.isArray(value) && value.length === 2;
+}
+
+function getValueForAxis<T = unknown>(
+  value: T | ForAxes<T> | undefined,
+  axis: Axis
+): T | undefined {
+  if (hasValueForAxis<T>(value, axis)) {
+    return value[axis] as T;
+  }
+  return value;
+}
```

Comment:
These will probably end up being shared helpers, since I see this logic repeated in a lot of places. Many victory props can be passed in as either `T` or `{ x: T, y: T }`, and these helpers provide some type safety for these props.